src/software/amazon/ion/impl/bin/IonRawBinaryWriter.java

```diff
@@ -1035,9 +1036,17 @@ public void writeInt(BigInteger value) throws IOException
     public void writeFloat(final double value) throws IOException
     {
         prepareValue();
-        updateLength(9);
-        buffer.writeUInt8(FLOAT_TYPE);
-        buffer.writeUInt64(doubleToRawLongBits(value));
+
+        if (value == ((double) ((float) value))) {
```

Comment:
It might be a good idea to make this behavior configurable.  `binary32` support was added as a relatively late feature in Ion and to maximize compatibility with other clients we should make this an opt-in.

The other ancillary benefit to opt-in is avoiding the narrow/widening/comparison machinery needed to do this which could have (probably marginal) performance implication.